File: typescript/0098-validate-binary-search-tree.ts
function isValidBST(root: TreeNode | null): boolean {
    return validate(root, null, null);
}

function validate(
    root: TreeNode | null,
    max: number | null,
    min: number | null
): boolean {
    if (!root) {
        return true;
    }

    if (
        (max !== null && root.val >= max) ||
        (min !== null && root.val <= min)
    ) {
        return false;
    }

    return (
        validate(root.left, root.val, min) &&
        validate(root.right, max, root.val)
    );
}

File: c/0098-validate-binary-search-tree.c
#define MAX 9223372036854775807
#define MIN -9223372036854775808

bool isSubTreeValid(struct TreeNode* root, long int lowerBound, long int upperBound) {
    if (root==NULL)
        return true;
    return root->val<upperBound && root->val>lowerBound && isSubTreeValid(root->left, lowerBound, root->val) && isSubTreeValid(root->right, root->val, upperBound);
}

bool isValidBST(struct TreeNode* root){
    return isSubTreeValid(root, MIN, MAX);
}
